Fix: ui_cluster: Improve crm cluster run command(bsc#1145520)

liangxin1300 opened the pull request:
**Assumption:** Two nodes cluster(15sp1-1 and 15sp1-2)

---
**Problem:** For a running two nodes cluster, run `crm cluster run "ls"`, will give this error:
```
ERROR: [15sp1-1]: Exited with error code 255, Error output: The authenticity of host '15sp1-1 (10.10.10.51)' can't be established.
ECDSA key fingerprint is SHA256:gOm0UDlChIQvNXXWn0N+bzVa+yj5Y80krR/i0boe/Z8.
Are you sure you want to continue connecting (yes/no)? 
Host key verification failed.
```
**Suggested Solution:** Add `'StrictHostKeyChecking=no'` ssh_options for parallax
**Related commit:** 2f84449238f0212958813bf5ab5584f75ff47e3b

---
**Problem:** When cluster service is totally stopped, run `crm cluster run "ls" 15sp1-1`, crmsh crashed:
```
ERROR: running cibadmin -Ql: Connection to the CIB manager failed: Transport endpoint is not connected
Init failed, could not perform requested operations
Traceback (most recent call last):
  File "/usr/sbin/crm", line 46, in <module>
    rc = main.run()
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 369, in run
    return main_input_loop(context, user_args)
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 249, in main_input_loop
    rc = handle_noninteractive_use(context, user_args)
  File "/usr/lib/python3.6/site-packages/crmsh/main.py", line 205, in handle_noninteractive_use
    if context.run(' '.join(l)):
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 86, in run
    rv = self.execute_command() is not False
  File "/usr/lib/python3.6/site-packages/crmsh/ui_context.py", line 276, in execute_command
    rv = self.command_info.function(*arglist)
  File "/usr/lib/python3.6/site-packages/crmsh/ui_cluster.py", line 610, in do_run
    if node and node in hosts:
TypeError: argument of type 'NoneType' is not iterable
```
**Reason:** That is because function `list_cluster_nodes` trying to use `cibadmin` to get the node list from cib.xml, and return `None` when failed, while we miss handling `None` result here

**Suggested Solution:**
* Changing the implementation of function `list_cluster_nodes`:
   - when pacemaker running, using `crm_node` to fetch node list
   - when corosync running, using `corosync-cmapctl` to fetch the member IP address
   - static situation, get env  CIB_file or parse file `/var/lib/pacemaker/cib/cib.xml` to get the node list
* Handle the `None` situation, give error message like `failed to get node list from cluster`

**Related commit:** 0dfa747bd9db3235990f5a6510dfd21ca55597c0

---

**Problem:** Can not run command for non-cluster node
**Suggested Solution:** If specific node in arguments, no matter whether it's a cluster node, enable running command on this node(s); Should also test whether need password, if yes, set `askpass` to `True` for parallax options
**Related commit:** f3cd676438f16b1ae2c96bdcb10194da439332df

krig commented:
Maybe you could split the change into two commits, first the update to `list_cluster_nodes` and another for changing the commands to take multiple nodes? This would make the commit messages clearer too :)

liangxin1300 commented:
Hi @krig @diegoakechi , please review this
And I also created commit e86b8548113a0d4a3385819d5076709be037ebfc to update doc

diegoakechi commented on crmsh/utils.py:
```diff
@@ -2166,4 +2176,26 @@ def debug_timestamp():
+    return ip_list
+
+
+def test_ssh_need_passwd(hosts):
```
@liangxin1300 I would avoid to use `test` prefix on a real function as it can cause confusion with unit tests. I would suggest here `check_ssh_passwd_need`.

liangxin1300 commented:
> @liangxin1300 I requested a small change.
> 
> Also, please, avoid to use only the Bug number on the commit title. Please, use more meaning description of what the real change is, as not all the bug information is public available.
> 
> I would like also to see some unit tests for these new scenarios. Can you, please, include it?
> 
> Thanks.

@diegoakechi Thanks for your suggestions, I've changed the function name `test_ssh_need_passwd` as `check_ssh_passwd_need`, changed the commit messages, and append unittest code
Please review
Thanks!:)

diegoakechi commented on test/unittests/test_utils.py:
```diff
@@ -12,6 +12,58 @@
+def test_check_ssh_passwd_need():
+    with mock.patch('crmsh.utils.get_stdout_stderr') as mock_get_stdout_stderr:
+        mock_get_stdout_stderr.side_effect = [(0, None, None), (1, None, None)]
+        assert utils.check_ssh_passwd_need(["node1", "node2"]) == True
```
@liangxin1300 we should also assert a False return case too.

liangxin1300 replied:
Yes, I defined this `def test_check_ssh_passwd_need_Flase():`

diegoakechi commented on test/unittests/test_utils.py:
```diff
@@ -12,6 +12,58 @@
+    mock_get_stdout_stderr.assert_called_once_with('corosync-cmapctl -b runtime.totem.pg.mrp.srp.members')
+
+
+def test_get_member_iplist():
```
@liangxin1300 We should have a test case for the case when the command failst too.

liangxin1300 replied:
@diegoakechi we have this function to test the command failed: `def test_get_member_iplist_None`

liangxin1300 commented:
@diegoakechi , please consider to merge this, cause it became a blocker for me when working on #464 
Thanks!
